ARCs/arc-0037.md

```diff
@@ -0,0 +1,77 @@
+---
+arc: 37
+title: ABI Versioning
+description: Add a versioning scheme to the ABI
+author: Joe Polny (@joe-p)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/210
+status: Draft
+type: Standards Track
+category: ARC
+created: 2023-07-15
+requires: 4
+---
+
+## Abstract
+The Algorand ABI is defined in [ARC-0004](./arc-0004.md). As the protocol and desires of the ABI changes, there will be changes to the ABI. As such, there should be a versioning number to make it clear what versions of the ABI smart contracts adhere to and what versions of the ABI off-chain code supports.
+
+## Motivation
+Without any versioning scheme, it will be very hard to follow the evolution of the Algorand ABI.
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+Every ABI change is versioned `MAJOR.MINOR`. `MAJOR` is incremeneted when there are breaking changes and `MINOR` is incremented for the addition of backwards-compatible features.
```

Comment:
I think it would be very useful to provide some examples of what constitutes a major vs minor update. Here's what I came up with so far.

#### Major
* Encoding changes
* Calling convention changes

All of these are certainly backwards incompatible and require the caller and callee to support the changes.

#### Minor:
* New types

Assume version 1.1 introduce a new type, `X`. Version 1.0 clients can still call methods on a version 1.1 contract, as long as those methods don't use `X` as an input or output type. So I _think_ this is still backwards compatible.